consensus/istanbul/backend/engine.go

```diff
@@ -175,20 +177,112 @@ func (sb *Backend) verifyCascadingFields(chain consensus.ChainReader, header *ty
 func (sb *Backend) VerifyHeaders(chain consensus.ChainReader, headers []*types.Header, seals []bool) (chan<- struct{}, <-chan error) {
 	abort := make(chan struct{})
 	results := make(chan error, len(headers))
-	go func() {
-		for i, header := range headers {
-			err := sb.verifyHeader(chain, header, headers[:i])
 
-			select {
-			case <-abort:
-				return
-			case results <- err:
+	// If we are in ultralight mode, we do batch verification of all BLS signatures
+	// in the headers, so there is no need to iterate over the blocks
+	if chain.Config().FullHeaderChainAvailable {
+		go func() {
+			for i, header := range headers {
+				err := sb.verifyHeader(chain, header, headers[:i])
+
+				select {
+				case <-abort:
+					return
+				case results <- err:
+				}
 			}
-		}
-	}()
+		}()
+	} else {
+		go func() {
+			err := sb.batchVerifyHeaders(chain, headers)
+			// even though we get 1 return value from the batched verification
+			// method, we need to fill the whole channel of results with return values
+			for i := 0; i < len(headers); i++ {
+				select {
+				case <-abort:
+					return
+				case results <- err:
+				}
+			}
+		}()
+	}
 	return abort, results
 }
 
+// batchVerifyHeaders is called in ultralight mode and performs batch BLS signature
+// verification over all provided headers and their corresponding validator sets
+func (sb *Backend) batchVerifyHeaders(chain consensus.ChainReader, headers []*types.Header) error {
+	// TODO: Populating this array can easily be done in parallel since the order does not matter
+	var msgs []*bls_ffi.SignedBlockHeader
+	for i, header := range headers {
+		// skip the genesis block
+		if header.Number.Uint64() == 0 {
+			continue
+		}
+
+		extra, err := types.ExtractIstanbulExtra(header)
+		if err != nil {
+			return err
+		}
+
+		// Don't waste time checking blocks from the future (allow some minor clock skew)
+		if header.Time > uint64(now().Unix())+mobileAllowedClockSkew {
+			return consensus.ErrFutureBlock
+		}
+
+		// short-circuit if there was no signature found
+		if len(extra.AggregatedSeal.Signature) == 0 {
+			return errEmptyAggregatedSeal
+		}
+
+		// 1. Get the aggregate public key for this epoch
+
+		// 1.1. Get the snapshot
+		snap, err := sb.snapshot(chain, header.Number.Uint64()-1, header.ParentHash, headers[:i])
```

Comment:
The `snapshot` function assumes that all of the headers passed in the `parents` parameter are already verified.   Does that assumption break anything here?

Reply:
The `batchVerifyHeaders` function is called with _all_ the headers which are provided to `engine.VerifyHeaders`. It then gathers all the signatures and validator sets from each epoch and forwards them to the signature verification algorithm. The verification algorithm _will_ fail if a "fake" validator public key is provided, if the signature is made by a different key.


 
Finally, my understanding is that the `parents` parameter is used as a cache, so if this is still a concern we can simply leave it as `nil`.

Reply:
Hmm yeah I'm not totally sure but I think @kevjue is onto something, nice catch.

When calling `snapshot` on header `H` in epoch N, it uses `parents` to build the validator set snapshot for `N` (and previous epochs, if necessary) by looking at the validator set diffs in `parents`.

A malicious full node could send a batch of headers that includes an invalid header with a validator set diff that changes out the validator set to keys controlled by the attacker. This batch of headers would get rejected, but not before calling `snapshot` on them, which calls `apply`, which would store the attacker's validator set. Then, the same full node could send another batch of headers signed by the attacker, which would get accepted.

Instead, we may need to pull a snapshot for the first header to get the validator set at the beginning of the batch, and then "manually" update this validator set whenever we see epoch blocks as we iterate over the headers.

Finally, if all checks pass and the signatures look good, then we call `snapshot` on the last header in the batch, passing all the parent headers, to update the validator set snapshot to the latest epoch.